src/v/redpanda/monitor_unsafe_log_flag.cc

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "redpanda/monitor_unsafe_log_flag.h"
+
+#include "cluster/logger.h"
+#include "utils/utf8.h"
+
+using cluster::clusterlog;
+
+class monitor_unsafe_log_flag::handle_unsafe_log_flag_change {
+public:
+    explicit handle_unsafe_log_flag_change(
+      cluster::cluster_version original_version)
+      : _original_version(original_version)
+      , _legacy_permit_unsafe_log_operation(
+          config::shard_local_cfg().legacy_permit_unsafe_log_operation.bind()) {
+        handle_flag_change();
+        _legacy_permit_unsafe_log_operation.watch(
+          [this] { handle_flag_change(); });
+    }
+
+private:
+    /*
+     * How the logic works:
+     *
+     * +------------------+------------+---------------------+
+     * | Original Version | Flag Value | Permit Unsafe Chars |
+     * +------------------+------------+---------------------+
+     * | >=v23.2.1        | X          | No                  |
+     * | <v23.2.1         | T          | Yes                 |
+     * | <v23.2.1         | F          | No                  |
+     * +------------------+------------+---------------------+
+     *
+     * If this is a new Redpanda cluster, then the value of the legacy flag
+     * is ignored.  If this is an upgraded cluster, then the legacy flag will
+     * control whether or not we permit control characters.
+     */
+    void handle_flag_change() {
+        if (_original_version >= flag_introduction_version) {
+            permit_unsafe_log_operation::set(false);
+        } else {
+            permit_unsafe_log_operation::set(
+              _legacy_permit_unsafe_log_operation());
+        }
+    }
+
+    cluster::cluster_version _original_version;
+    config::binding<bool> _legacy_permit_unsafe_log_operation;
+};
+
+monitor_unsafe_log_flag::monitor_unsafe_log_flag(
+  ss::abort_source& shard0_as, cluster::cluster_version original_version)
+  : _shard0_as(shard0_as)
+  , _original_version(original_version) {}
+
+ss::future<> monitor_unsafe_log_flag::maybe_log_flag_nag() {
+    auto nag_check_retry = std::chrono::seconds{300};
```

Comment:
should this be tunable? i could see someone wanting to silence the nag?

Reply:
I'm mimicking what's in `feature_manager`: https://github.com/redpanda-data/redpanda/blob/8a18ba77955311dbad6d9ff4ac04a94e1a097deb/src/v/cluster/feature_manager.cc#L180-L229

But i could make it tunable for sure

Reply:
yeh. it's a curious thing: if i explicitly set the option to allow unsafe in legacy systems, am i also defacto opting out of the need for a nag?

Reply:
Well in legacy systems the flag is default on (permit unsafe) so the operator is not making an active decision to enable the flag.  So I think the nag is warranted.

Reply:
One other thing to add is that within a couple of releases all of this code will be removed.  Would you still want the nag check to be tunable?  I can add in a new configuration flag if so desired.

Reply:
Maybe we can not run this on a timer? And instead just print it when the option is set/unset?

Reply:
Well it's set to default to true (permit unsafe) on upgrades so unless the customer is purposefully looking at the start of the logs, they may miss it.  I feel being somewhat 'obnoxious' would be helpful to communicate to the operator that there may be an issue in the future.

Reply:
Maybe this is more of a question for product then

Reply:
I'll double check with them, but I do think this came from them.  (DOn't have a PRD to reference, and I need to update the RFC).  This came from quick meeting I had with product/CS

Reply:
>Maybe we can not run this on a timer? And instead just print it when the option is set/unset?

this would be ideal. it'd be nice to not have a timer and added complexity just for the nag message. we could even imagine bubbling it up into the health reports or status etc... at some point.

Reply:
I'll re-ask. I guess I don't understand the concern about the "complexity". This is very similar to how the license nag works.

Reply:
Talked with @dcoded about this and there is a desire to have the nag